2022/Q2/artifacts/pos-model/Staking.tla

```diff
@@ -0,0 +1,179 @@
+        a \in UserAddrs |-> IF a /= "validator"
+        THEN 0
+        ELSE INIT_VALIDATOR_STAKE
+       ]
+    /\ nextTxId = 0
+    /\ epoch = 1
+    /\ txCounter = 0
+    /\ lastTx = [ id |-> 0, tag |-> "None", fail |-> FALSE ]
+    /\ failed = FALSE
+
+
+\* delegate `amount` coins to Validator
+Delegate(sender, amount) ==
+    LET fail ==
+        \/ amount < 0
+        \/ amount > balanceOf[sender]
+    IN
+    /\ lastTx' = [ id |-> nextTxId, tag |-> "delegate", fail |-> fail,
+                   sender |-> sender, toAddr |-> Validator, value |-> amount ]
+    /\ nextTxId' = nextTxId + 1
+    /\ failed' = (fail \/ failed)
+    /\  IF fail
+        THEN
+          UNCHANGED <<balanceOf, unbonded, delegated>>
+        ELSE
+          \* transaction succeeds
+          \* update the balance of 'sender'
+          /\ balanceOf' =
+                [ balanceOf EXCEPT ![sender] = @ - amount]
+          /\ delegated' = [ delegated EXCEPT ![sender] = @ + amount ]
+          /\ UNCHANGED unbonded
+
+
+\* unbond `amount` coins from Validator
+Unbond(sender, amount) ==
+    LET fail ==
+        \/ amount < 0
+        \/ sender = Validator
+        \/ amount > delegated[sender]
+    IN
+    /\ lastTx' = [ id |-> nextTxId, tag |-> "unbond", fail |-> fail,
+                   sender |-> sender, toAddr |-> Validator, value |-> amount ]
+    /\ nextTxId' = nextTxId + 1
+    /\ failed' = (fail \/ failed)
+    /\  IF fail
+        THEN
+          UNCHANGED <<balanceOf, unbonded, delegated>>
+        ELSE
+          \* transaction succeeds
+          /\ unbonded' = [ unbonded EXCEPT ![sender] = @ + amount ]
+          /\ delegated' = [ delegated EXCEPT ![sender] = @ - amount ]
+          /\ UNCHANGED  balanceOf
+
+\* withdraw unbonded coins
+Withdraw(sender) ==
+    LET fail ==
+        \/ sender = Validator
+        \/ unbonded[sender] > 0
+    IN
+    /\ lastTx' = [ id |-> nextTxId, tag |-> "withdraw", fail |-> fail,
+                   sender |-> sender, toAddr |-> Validator, value |-> unbonded[sender] ]
+    /\ nextTxId' = nextTxId + 1
+    /\ failed' = (fail \/ failed)
+    /\  IF fail
+        THEN
+          UNCHANGED <<balanceOf, unbonded, delegated>>
+        ELSE
+          \* transaction succeeds
+          /\ balanceOf' = [ balanceOf EXCEPT ![sender] = @ + unbonded[sender] ]
+          /\ unbonded' = [ unbonded EXCEPT ![sender] = 0 ]
+          /\ UNCHANGED  delegated
+
+AdvanceEpoch ==
+    /\  IF txCounter = TXS_PER_EPOCH
+        THEN
+          \* move to the next epoch
+          /\ epoch' = epoch + 1
+          /\ txCounter' = 0
+        ELSE
+          \* do not advance epoch
+          /\ txCounter' = txCounter + 1
+          /\ UNCHANGED  epoch
+
+\* The transition relation, which chooses one of the actions
+Next ==
+    \/ \E sender \in UserAddrs:
+       \E amount \in Int:
+        /\ amount <= MAX_UINT
+        /\ \/ Delegate(sender, amount)
+           \/ Unbond(sender, amount)
+           \/ Withdraw(sender)
+        /\ AdvanceEpoch
```

Comment:
(related to the question above) How do you ensure that "AdvanceEpoch" is interleaved one-to-one with one transaction?

Reply:
Since we have `/\` in lines 166-170, `AdvanceEpoch` is effectively done in lock step with one of the actions: `Delegate`, `Unbond`, or `Withdraw`.

Reply:
Of course! I get it now